enhancements/storage/storage-hypershift.md

```diff
@@ -0,0 +1,1050 @@
+    * Rework CSI driver Deployment to run only the control plane parts of the CSI driver, i.e.:
+      * Remove `hostNetwork: true`.
+      * Add `controller` as the first command line arg.
+      * Set env. var. `AWS_REGION=us-east2` with the actual guest cluster region. The region can be read from
+        `Infrastructure` CR.
+      * This removes dependency on AWS metadata service and allows us to run the controller parts of the driver outside
+        of AWS.
+    * Rework CSI driver DaemonSet to run only node parts of the CSI driver.
+      * Add `node` as the first command line arg.
+      * This allows us not to give the DaemonSet any AWS credentials.
+  * Only on HyperShift:
+    * When interacting with Kubernetes API, carefully choose the right kubeconfig to watch / create / update objects in
+      the right cluster.
+    * Replace namespaces in the driver Deployment, PDBs and other objects that are created in the management cluster.
+      They must be created in the same namespace as the operator.
+    * Update the CSI driver control-plane Deployment: 
+      * Give it kubeconfig to the guest cluster as a Secret volume.
+      * Pass it Secret with STS creds generated by control-plane-operator.
+      * Add `token-minter` sidecar, which will create STS secrets to the driver.
+      * Set `priorityClassName: hypershift-control-plane`.
+    * Phase 2: add metrics services + TLS keys.
+
+* **AWS EBS CSI driver**:
+  * Phase 1: no code changes.
+    * Controller parts will run in the management cluster, but they get kubeconfig to the guest cluster.
+    * Node DaemonSet will get the default projected kubeconfig that can access objects only in the
+    openshift-cluster-csi-drivers namespace.
+  * Phase 2 (or even later): reduce memory requirements of AWS EBS CSI driver control-plane Deployment.
+    * Review memory requirements of each sidecar.
+    * Union kube-rbac-proxies, if possible.
+    * Union all sidecars into a single container, so they share informers? This will need serious investment upstream.
+  * Phase 2 (or even later): move leader-election of CSI sidecars to the management cluster.
+    * I.e. the sidecars get a special kubeconfig just for the leader election.
+    * The sidecars already have a parameter with namespace where the Lease objects are created, CSI driver operators
+      will need to set correctly in HyperShift, so CSI sidecars of different guest clusters don't collide.
+
+#### Notes about other CSI drivers
+
+* Azure Disk CSI driver
+  * It needs some cloud credentials available on nodes. We need to check what permissions the CSI driver pods that run
+    on worker nodes actually need and perhaps create two roles in Azure - one for the control-plane of the CSI driver
+    (create / delete / attach / detach volumes) and second for the node (read volume details?). 
+    
+* vSphere CSI driver
+  * Controller parts of the CSI driver assume that they run in vSphere. Significant driver changes may be needed
```

Comment:
Are they using some sort of interface to talk to vSphere that only works if you're actually inside the vSphere cluster (device driver, etc.)? Or do they just need an API endpoint and credentials?

Reply:
@gnufied, can you please provide details why the driver needs to run in vSphere even for control plane pods?

Reply:
Discussed offline, it seems that the controller pods don't need to run in vSphere. I removed the bullet.